logexp/logexp.sh

```diff
@@ -0,0 +1,148 @@
+#!/bin/bash
+# Copyright 2017 The Kubernetes Authors.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# This script when run on a node (/master) in a kuberentes cluster would
+# export the logs to gcs, with the upload path and other config taken
+# through environment variables.
+
+set -o errexit
+set -o nounset
+set -o pipefail
+
+# The following env vars need to be defined before calling this script:
```

Comment:
The flag and golang suggestion are separate.

Flags make it easier to trace dependencies and predict the behavior of the program based on how the program is launched vs global environment state where it was launched.

Golang makes it easier to predict the impact of changes. It is also easier to write unit tests (please consider writing some unit tests to cover interesting and/or non-trivial scenarios):

* program exits if the service account creds are not specified
* we add the right files based on the provider
* we do not crash if journalctl fails
* we retry gsutil uploads that flake
* we detect systemd vs supervisord correctly
* etc